### PR TITLE
version: improve alpha and rc detection (elasticsearch)

### PR DESCRIPTION
**Pull request by zmwangx**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Previously we were in a weird situation where `elasticsearch 5.0.0-beta1` can be detected but `5.0.0-alpha5` cannot.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
While you're playing with this might want to handle the openssl case @DomT4 mentioned?

**Comment by zmwangx** [comment]
@MikeMcQuaid We seem to be parsing openssl versions just fine? The version comparison issue is handled by `version_scheme`.

**Comment by MikeMcQuaid** [comment]
I believe `1.0.0a` is interpreted as an alpha, incorrectly.

**Comment by zmwangx** [comment]
That's the expected behavior due to

```ruby
  class AlphaToken < CompositeToken
    PATTERN = /a(?:lpha)?[0-9]*/i

    def <=>(other)
      case other
      when AlphaToken
	rev <=> other.rev
      else
        super
      end
    end
  end
```

We might be able to change that pattern to `/a[0-9]+|alpha[0-9]*/`, but I'm not sure if that's going to break any existing formula.

**Comment by MikeMcQuaid** [comment]
> We might be able to change that pattern to /a[0-9]+|alpha[0-9]*/, but I'm not sure if that's going to break any existing formula.

I think in this case given we know it breaks at least one pretty popular formula as-is it's worth making the change. @DomT4, thoughts?

**Comment by zmwangx** [comment]
By the way I noticed that non-capturing groups are used all over the place in `versions.rb`. What's the point?

File: Library/Homebrew/version.rb [hunk 1]
```diff
@@ -340,8 +340,8 @@ def self._parse(spec)
     m = /-((?:\d+\.)*\d+(?:[abc]|rc|RC)\d*)$/.match(stem)
     return m.captures.first unless m.nil?
 
-    # e.g. foobar-4.5.0-beta1, or foobar-4.50-beta
-    m = /-((?:\d+\.)*\d+-beta\d*)$/.match(stem)
+    # e.g. foobar-4.5.0-alpha5, foobar-4.5.0-beta1, or foobar-4.50-beta
+    m = /-((?:\d+\.)*\d+-(alpha|beta|rc)\d*)$/.match(stem)
```
Comment (zmwangx) [inline Library/Homebrew/version.rb]: Might want to use a non-capturing group here to be consistent with everything else, but I want to know why they're used in the first place.

**Comment by MikeMcQuaid** [comment]
> What's the point (when we only need .captures.first anyway)?

Not sure. Could be it used to use multiple captures.

**Comment by zmwangx** [comment]
Do we want to keep using non-capturing groups, or remove them where unnecessary en masse?

**Comment by MikeMcQuaid** [comment]
I don't mind, up to you.

**Comment by zmwangx** [comment]
I'll keep using non-capturing group here for consistency, and maybe open a separate PR afterwards to handle the removal.

**Comment by DomT4** [comment]
> I think in this case given we know it breaks at least one pretty popular formula as-is it's worth making the change. @DomT4, thoughts?

I'm not too fussed about going out of our way to handle OpenSSL. They stand out as having a particularly unusual version scheme & it's not necessary a surprise that doesn't mesh well with our overall version logic in places. Everything else seems to be more or less working as intended/

If people want to tweak it & test that against a bunch of places where we use `a` in the version & that `a` is no longer detected as an alpha releases I'd be fine with that, but I don't think the status quo is a massive problem outside of OpenSSL's version weirdness.

**Comment by zmwangx** [comment]
Maybe we should just override version code inside openssl formulae themselves?

**Comment by MikeMcQuaid** [comment]
> I'm not too fussed about going out of our way to handle OpenSSL. They stand out as having a particularly unusual version scheme & it's not necessary a surprise that doesn't mesh well with our overall version logic in places. Everything else seems to be more or less working as intended/

I don't know if I agree; having `a` on the end of a version doesn't seem to imply alpha to me.

**Comment by zmwangx** [comment]
I'm merging this because the openssl discussion isn't directly related to the fairly straightforward change in this PR. Please keep the discussion going if you will, though.

**Comment by chdiza** [comment]
There are some other things that misleadingly put an "a" on the end of non-alpha versions at times.  I think at least one of the `gmp`/`mpfr` duo has done that.  I also vaguely recall `tmux` doing that at least once.  Maybe there should be a hash that holds all the formulas at a given time that are like this, and use it has a kind of whitelist against accusations of alphahood.

**Pull request merged**